src/java/htsjdk/samtools/sra/SRAIndexedSequenceFile.java

```diff
@@ -0,0 +1,138 @@
+    private Map<ReferenceCacheKey, Reference> cachedReferencesByThreads = new ConcurrentHashMap<ReferenceCacheKey, Reference>();
+
+    private Iterator<SAMSequenceRecord> sequenceRecordIterator;
+
+    protected SAMSequenceDictionary sequenceDictionary;
+
+    private static class ReferenceCacheKey {
+        long threadId;
+        int sequenceIndex;
+
+        ReferenceCacheKey(long threadId, int sequenceIndex) {
+            this.threadId = threadId;
+            this.sequenceIndex = sequenceIndex;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj)
+                return true;
+            if (obj == null)
+                return false;
+            if (getClass() != obj.getClass())
+                return false;
+            ReferenceCacheKey other = (ReferenceCacheKey) obj;
+
+            return other.threadId == this.threadId && other.sequenceIndex == this.sequenceIndex;
+        }
+
+        @Override
+        public int hashCode() {
+            return (int) (threadId) ^ sequenceIndex;
+        }
+    }
+
+    public SRAIndexedSequenceFile(SRAAccession acc) {
+        this.acc = acc;
+
+        if (!acc.isValid()) {
+            throw new RuntimeException("Passed an invalid SRA accession into SRA reader: " + acc);
+        }
+
+        try {
+            run = gov.nih.nlm.ncbi.ngs.NGS.openReadCollection(acc.toString());
+            sequenceDictionary = loadSequenceDictionary();
+        } catch (final ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+
+        reset();
+    }
+
+    @Override
+    public SAMSequenceDictionary getSequenceDictionary() {
+        return sequenceDictionary;
+    }
+
+    @Override
+    public ReferenceSequence nextSequence() {
+        SAMSequenceRecord sequence = sequenceRecordIterator.next();
+        return getSubsequenceAt(sequence.getSequenceName(), 1L, sequence.getSequenceLength());
+    }
+
+    @Override
+    public void reset() {
+        sequenceRecordIterator = sequenceDictionary.getSequences().iterator();
+    }
+
+    @Override
+    public boolean isIndexed() {
+        return true;
+    }
+
+    @Override
+    public ReferenceSequence getSequence(String contig) {
+        return getSubsequenceAt(contig, 1L, sequenceDictionary.getSequence(contig).getSequenceLength());
+    }
+
+    @Override
+    public ReferenceSequence getSubsequenceAt(String contig, long start, long stop) {
+        SAMSequenceRecord sequence = sequenceDictionary.getSequence(contig);
+        int referenceIndex = sequence.getSequenceIndex();
+
+        byte[] bases;
+
+        try {
+            ReferenceCacheKey key = new ReferenceCacheKey(Thread.currentThread().getId(), referenceIndex);
+            Reference reference = cachedReferencesByThreads.get(key);
+            if (reference == null) {
+                reference = run.getReference(contig);
+                cachedReferencesByThreads.put(key, reference);
+            }
```

Comment:
Why mess around with thread IDs manually like this? Why not just make the cache itself `ThreadLocal`, which would be much simpler and less error-prone?